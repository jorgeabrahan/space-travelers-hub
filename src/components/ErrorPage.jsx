import styles from '../css/Error.module.css';

function ErrorPage() {
  return (
    <div className={styles.error}>
      <h2 className={styles.h2}>Ups...Something went wrong!</h2>
      <p className={styles.p}>
        404 not found
      </p>
    </div>
  );
}

export default ErrorPage;
