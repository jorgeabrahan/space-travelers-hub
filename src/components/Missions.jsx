import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { missionsFetchAPI } from '../redux/missions/missionsSlice';
import Mission from './Mission';
import styles from '../css/Missions.module.css';

function Missions() {
  const dispatch = useDispatch();
  const { missions, status } = useSelector((store) => store.missions);

  useEffect(() => {
    if (missions.length > 0) return;
    if (status === 'failed') return;
    dispatch(missionsFetchAPI());
  }, [dispatch]);

  if (status !== 'succeeded') {
    return (
      <section className="center">loading...</section>
    );
  }
  return (
    <section className="center">
      <table className={styles.table}>
        <thead>
          <tr>
            <th className={styles.tableHeader}>Mission</th>
            <th className={styles.tableHeader}>Description</th>
            <th className={styles.tableHeader}>Status</th>
            <th className={styles.tableHeader}> </th>
          </tr>
        </thead>
        <tbody>
          {
            missions.missions.map((item, i) => (
              <Mission
                key={item.mission_id}
                missionName={item.mission_name}
                description={item.description}
                odd={i % 2 === 0}
              />
            ))
          }
        </tbody>
      </table>
    </section>
  );
}

export default Missions;
