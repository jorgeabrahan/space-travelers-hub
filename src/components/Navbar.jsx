import { NavLink, Outlet } from 'react-router-dom';
import planet from '../planet.png';
import styles from '../css/Navbar.module.css';

function Navbar() {
  return (
    <>
      <header>
        <nav className={styles.nav}>
          <div className={styles.column}>
            <img className={styles.logo} src={planet} alt="Logo" />
            <h1 className={styles.h1}>Space Travelers&apos; Hub</h1>
          </div>
          <div className={styles.column}>
            <ul className={styles.ul}>
              <li className={styles.li}>
                <NavLink
                  to="rockets"
                  className={({ isActive }) => {
                    const active = isActive ? styles.active : styles.navLink;
                    return active;
                  }}
                >
                  Rockets
                </NavLink>
              </li>
              <li className={styles.li}>
                <NavLink
                  to="/"
                  className={({ isActive }) => {
                    const active = isActive ? styles.active : styles.navLink;
                    return active;
                  }}
                >
                  Missions
                </NavLink>
              </li>
              <li className={styles.li}>
                <NavLink
                  to="profile"
                  className={({ isActive }) => {
                    const active = isActive ? styles.active : styles.navLink;
                    return active;
                  }}
                >
                  Profile
                </NavLink>
              </li>
            </ul>
          </div>
        </nav>
      </header>
      <Outlet />
    </>
  );
}

export default Navbar;
