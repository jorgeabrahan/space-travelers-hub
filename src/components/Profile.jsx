function Profile() {
  return (
    <section className="center">
      <h1>This is Profile Component</h1>
    </section>
  );
}

export default Profile;
