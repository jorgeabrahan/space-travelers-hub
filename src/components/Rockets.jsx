import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets } from '../redux/rockets/rocketsSlice';
import Rocket from './Rocket';

function Rockets() {
  const dispatch = useDispatch();
  const { rockets, status } = useSelector((store) => store.rockets);
  useEffect(() => {
    /* Validation to fetch data just when the page loads */
    if (rockets.length > 0) return;
    if (status === 'failed') return;
    dispatch(fetchRockets());
  }, [dispatch]);

  const rocketsList = rockets.map(({
    id, name, type, images, description,
  }) => (
    <Rocket
      key={id}
      name={name}
      type={type}
      image={images[0]}
      description={description}
    />
  ));

  return (
    <section className="center">
      {rocketsList}
    </section>
  );
}

export default Rockets;
