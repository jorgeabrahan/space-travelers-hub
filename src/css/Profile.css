.profile {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  gap: 1rem;
}

.reserved__title {
  font-size: 1.8rem;
  margin-bottom: 0.5rem;
}

.reserved__list {
  border: 2px solid #e2e2e2;
  border-radius: 5px;
}

.reserved__item {
  padding: 1rem 1.5rem;
  font-size: 1.2rem;
}

.reserved__item:not(:last-child) {
  border-bottom: 2px solid #e2e2e2;
}

.reserved__message {
  font-size: 1.2rem;
}
