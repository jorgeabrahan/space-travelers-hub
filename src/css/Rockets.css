.rockets {
  display: grid;
  gap: 1.5rem;
}

.rocket {
  display: flex;
  gap: 1rem;
}

.rocket__image {
  width: 260px;
}

.rocket__title {
  font-size: 1.4rem;
  margin-bottom: 0.3rem;
}

.rocket__badge {
  background-color: #18a2b8;
  color: white;
  font-size: 0.85rem;
  font-weight: bold;
  padding: 0.1rem 0.3rem;
  border-radius: 5px;
  margin-right: 0.6rem;
}

.rocket__description {
  font-size: 1.05rem;
  line-height: 1.5rem;
  font-weight: 400;
  margin-bottom: 0.8rem;
}

.rocket__button {
  border: none;
  border-radius: 5px;
  font-size: 1.1rem;
  padding: 0.5rem 1rem;
}

.rocket__button.reserve {
  background-color: #007bff;
  color: white;
}

.rocket__button.reserved {
  background: none;
  border: 1px solid #606972;
  color: #606972;
}
