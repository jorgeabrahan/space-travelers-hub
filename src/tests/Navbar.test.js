import { render, screen } from '@testing-library/react';
import { BrowserRouter, createBrowserRouter, RouterProvider } from 'react-router-dom';
import { configureStore, createSlice } from '@reduxjs/toolkit';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../components/Missions';
import Navbar from '../components/Navbar';
import Profile from '../components/Profile';
import Rockets from '../components/Rockets';
import ErrorPage from '../components/ErrorPage';

describe('Tests for <Navbar /> component', () => {
  const name = 'Mission 1';
  const description = 'Mission description';
  const initialState = {
    missions: [{
      mission_id: '1',
      mission_name: name,
      description,
      reserved: false,
    }, {
      mission_id: '2',
      mission_name: name,
      description,
      reserved: true,
    }],
    status: 'succeeded',
    error: '',
  };

  const title = 'Falcon 1';
  const titleNoSpace = title.replace(/\s/g, '');
  const descriptionR = 'Falcon description';
  const initialStateR = {
    rockets: [{
      id: '1',
      name: title,
      images: [`${titleNoSpace}.jpg`],
      description: descriptionR,
      reserved: false,
    }, {
      id: '2',
      name: title,
      images: [`${titleNoSpace}.jpg`],
      description: descriptionR,
      reserved: true,
    }],
    status: 'succeeded',
    error: '',
  };

  const rocketsSlice = createSlice({
    name: 'rockets',
    initialState: initialStateR,
  });

  const missionSlice = createSlice({
    name: 'missions',
    initialState,
  });

  const store = configureStore({
    reducer: {
      missions: missionSlice.reducer,
      rockets: rocketsSlice.reducer,
    },
  });

  const routerProvider = () => {
    const router = createBrowserRouter([
      {
        path: '/',
        element: <Navbar />,
        errorElement: <ErrorPage />,
        children: [
          {
            index: true,
            element: <Rockets />,
          },
          {
            path: 'missions',
            element: <Missions />,
          },
          {
            path: 'profile',
            element: <Profile />,
          },
        ],
      },
    ]);
    return (
      <Provider store={store}>
        <RouterProvider router={router} />
      </Provider>
    );
  };

  it('Testing Rockets route', () => {
    render(routerProvider(<Navbar />));
    // This is a text from Rockets component
    expect(screen.getByText('Rockets')).not.toBeNull();
  });

  it('Testing Missions route', () => {
    render(routerProvider(<Navbar />));
    // This is a text from Missions component
    expect(() => screen.getByText('Missions')).not.toBeNull();
  });

  it('Testing Profile route', () => {
    render(routerProvider(<Navbar />));
    // This is a text from Profile component
    expect(() => screen.getByText('Profile')).not.toBeNull();
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(<BrowserRouter className="app-container"><Navbar /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });
});
