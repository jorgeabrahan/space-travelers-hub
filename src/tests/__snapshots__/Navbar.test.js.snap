// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Navbar /> component Should match the snapshot 1`] = `
<header>
  <nav
    className="nav center"
  >
    <div
      className="column"
    >
      <img
        alt="Logo"
        className="logo"
        src="planet.png"
      />
      <h1
        className="h1"
      >
        Space Travelers' Hub
      </h1>
    </div>
    <div
      className="column"
    >
      <ul
        className="ul"
      >
        <li
          className="li"
        >
          <a
            aria-current="page"
            className="active"
            href="/"
            onClick={[Function]}
          >
            Rockets
          </a>
        </li>
        <li
          className="li"
        >
          <a
            className="navLink"
            href="/missions"
            onClick={[Function]}
          >
            Missions
          </a>
        </li>
        <li
          className="li"
        >
          <a
            className="navLink"
            href="/profile"
            onClick={[Function]}
          >
            Profile
          </a>
        </li>
      </ul>
    </div>
  </nav>
</header>
`;
